File: 2021/src/day1/day1.ts
import { DayConfig } from '../Day';
import { readMultilineNumberList } from '../read';

function countIncreasingSteps(input: number[]) {
  let counter = 0;
  let prev = input[0];
  for (let i = 1; i < input.length; i++) {
    const next = input[i];
    if (next > prev) {
      counter++;
    }
    prev = next;
  }
  return counter;
}

const part1 = (input: number[]): number => {
  return countIncreasingSteps(input);
};

const part2 = (input: number[]): number => {
  let windows = listToWindows(input);
  return countIncreasingSteps(windows);
};


const listToWindows = (input: number[]): number[] => {
  let windows = [input[0] + input[1] + input[3]];
  for (let i = 3; i < input.length; i++) {
    let nextNum = input[i];
    let removedNum = input[i - 3];
    let prevSum = windows[windows.length - 1];
    windows.push(prevSum + nextNum - removedNum);
  }
  return windows;
};


export const Day1: DayConfig<number[]> = {
  dayNumber: 1,
  parseInput: readMultilineNumberList,
  part1,
  part2,
};

File: cpp-2021/day1/day1.cpp
//
// Created by Devon D'Apuzzo on 5/5/22.
//

#include "day1.h"
#include <vector>

int Day1::Part1() {
  int count = 0;
  int prev = INT16_MAX;
  for (int i = 0; i < days.size(); ++i) {
    if (days[i] > prev) {
      count++;
    }
    prev = days[i];
  }
  return count;
}

int Day1::Part2() {
  int count = 0;
  int prev = days[0] + days[1] + days[2];

  for (int i = 3; i < days.size(); ++i) {
    int next = prev - days[i - 3] + days[i];

    if (next > prev) {
      count++;
    }
    prev = next;
  }
  return count;
}
Day1::Day1(const std::vector<int> &days) : days(days) {}
